### PR TITLE
OKTA Authentication swap to Auth0

### PR DESCRIPTION
**Pull request by da-vazquez**
# Pull Request Template

Replaced Okta embedded Login Form with a button component that routes the user to the Auth0 Universal Login page. Modified Api/index.js by removing authState (okta infrastructure) and replaced it with an idToken and Authentication boolean value pulled from localStorage that is then sent to the backend API for verification.

    1. User can login with their current email or password or sign-up on the same form (provided by Auth0)
    2. Auth0 form will be styled at a later time to match current SS look and feel
    
## Description
Motivation/purpose for the feature:
     1. To ensure the user can login / register and to replace the current Okta Authentication system currently in place. All of the features for SS require an authenticated user so this feature is critical to the user experience.
     2. Allows current architecture to remain functional using Auth0
 
- What were those changes
     1. A button was created on the LandingPage.js component that will direct the user to the Auth0 Login screen and re-direct back to the SS application after the credentials are verified.
     2. After the user logs in, the application stores the user's idToken and authenication status in localStorage so it can be consumed by the rest of the components
 
- What issues were fixed
     1. The embedded Okta sign-in widget was removed and replaced by a login button that directs to the Auth0 Login Page.
     2. Removed all instances of Okta on login component and parent dashboard component
     3. API/index.js removed all instances of authState and updated getAuthHeader to send idToken to API
     3. Tests updated to mock Auth0 instead of Okta
     4. Remaining components have also been modified on separate PR's and will be merged in multiple stages to ensure there are no bugs

 - Provide the link to the user story it is referencing in Trello board or other supporting document
     https://trello.com/c/uoBFFr9W/26-as-a-user-i-want-to-be-able-to-login-using-auth0-on-the-story-squad-website

## Notes

Additional modifications were made to ParentDashBoardContainer.js, ParentNavTopBar.js, and SecureRoute.js as they pertained to testing the Auth0 redirect functionality. Without the modifications, the application was crashing. Backend API Pull Request will be merged simultaneously with this PR.

## Commits Checklist
- [x] commits have clear title names 
- [x] commit has descriptive message of what has changed 
- [x] more small commits than less big commits 

## Code Checklist 
- [x] code is formatted appropriately 
- [x] code meets DRY standards
- [X] no dead code (check for logs and print statements) (a few commented out lines, will need additional clarification if code is needed or not)
- [x] no TODOS instead necessary comments 
- [x] code follows Story Squad standards for:
     - Language 
     - Framework 
     - Libraries   

### PRs needs to be reviewed by at least 2 team members
 - TPL is the 3rd reviewer 

[Pull Request Rubric](https://www.notion.so/1fc04e4fedeb429ba873b7c68d281707?v=74054da7991341c0bf970f39410c43da)

### CODE REVIEW

File: src/components/common/ParentNavTopBar.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 import { Typography, Menu, Dropdown, Button } from 'antd';
 import { Link, useHistory } from 'react-router-dom';
-import { useOktaAuth } from '@okta/okta-react';
+import { useAuth0 } from '@auth0/auth0-react'; //replaces okta
```
Comment (SethMacP) [inline src/components/common/ParentNavTopBar.js]: While I appreciate the comment, I don't believe it will need to stay in the codebase as Okta will not be around anymore. From the PR it's fairly self-explanatory as you see both the removed line and the added line with both saying they are an Auth tool.

File: src/components/common/SecureRouteAuth0.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
```
Comment (kdneldor) [inline src/components/common/SecureRouteAuth0.js:1]: Nice job making this component! It is nice and modular, making it easy to implement.

File: src/components/pages/AccountSettings/RenderAccountSettings.js [hunk 3]
```diff
@@ -1,29 +1,28 @@
 import React, { useState, useEffect } from 'react';
 import { Modal, Button, Form } from 'antd';
-import { useOktaAuth } from '@okta/okta-react';
+import { useAuth0 } from '@auth0/auth0-react'; //replaces okta
```
Comment (SethMacP) [inline src/components/pages/AccountSettings/RenderAccountSettings.js]: Same comment as above about the commented out message.

File: src/components/pages/NewParentDashboard/NewParentDashboardContainer.js [hunk 4]
```diff
@@ -1,45 +1,42 @@
...
 function NewParentDashboardContainer({ LoadingComponent, ...props }) {
-  const { authState, authService } = useOktaAuth();
-  const [userInfo, setUserInfo] = useState(null);
+  const { user, isAuthenticated } = useAuth0();
+  const [userInfo, setUserInfo] = useState(user);
   // eslint-disable-next-line
-  const [memoAuthService] = useMemo(() => [authService], []);
+  const [memoAuthService] = useMemo(() => [user], []);
+
+  /* do we need code below since auth0 has built-in isAuthenticated hook? */
```
Comment (Antashma) [inline src/components/pages/NewParentDashboard/NewParentDashboardContainer.js]: ❕I would ask this at the next stakeholder meeting or see if we can meet with Trevor again sooner to review

File: src/components/pages/AccountSettings/RenderAccountSettings.js (hunk 3 above)
Comment (da-vazquez) [reply src/components/pages/AccountSettings/RenderAccountSettings.js]: Great catch @SethMacP!

File: src/index.js [hunk 5]
```diff
@@ -16,13 +16,15 @@ import {
 } from 'react-router-dom';
 
 import { Security } from '@okta/okta-react';
```
Comment (Antashma) [inline src/index.js]: Can this line be removed since the line below replaces it?

File: src/components/pages/Login/LoginContainer.js [hunk 6]
```diff
@@ -1,55 +1,14 @@
-import React, { useEffect } from 'react';
```
Comment (kdneldor) [inline src/components/pages/Login/LoginContainer.js:1]: Nice job implementing the redirect functionality in this component! Since it is only has 49 lines that were removed (maybe that is a lot) I am not too worried about it being not atomic in nature.

File: src/index.js [hunk 7]
```diff
@@ -16,13 +16,15 @@ import {
...
+import { Auth0Provider } from '@auth0/auth0-react'; //this will replace okta above
```
Comment (SethMacP) [inline src/index.js]: You say 'this **will** replace Okta above', but I see it implemented further down in the code. Is there reasoning for keeping the old systems in place? From what I can tell you've swapped all of the secureRoutes to protectedRoutes in this file.

**Review by SethMacP — DISMISSED** [review]
Great work, Daniel. You managed to get a lot done in this PR and I appreciated that you left a comment about the usefulness of that commented out code. I'd advise following up with someone outside of this PR to get an answer quicker. 

My only 'non code' critique would be to check the length of your commits. The overwhelming majority seemed to have the '...' on them.

**Review by kdneldor — COMMENTED** [review]
Great work Daniel! You gave good details when it came to what is going on in this PR. Commits are well written and atomic. In general it seems appropriately scoped. Your motivation is clear. My main concern would be the amount of files changed, but 10 may not be a big deal.

**Comment by Antashma** [comment]
This looks great overall!  I'm just concerned about the large amount of commented out code that you may be able to remove entirely (please see the comments).  I also will look into that coverage test failure.

File: src/index.js (hunk 5 above)
Comment (da-vazquez) [reply src/index.js]: Absolutely! I will remove this line and the wrapper around the components as it has been replaced with Auth0 ProtectedRoute

File: src/components/pages/Login/LoginContainer.js (hunk 6 above)
Comment (da-vazquez) [reply src/components/pages/Login/LoginContainer.js:1]: Thank you @kdneldor, the Okta Widget was pretty heavy on code but Auth0 is very lightweight and only requires a few lines to get the job done :)

File: src/index.js (hunk 7 above)
Comment (da-vazquez) [reply src/index.js]: Absolutely! I will remove this line and the wrapper around the components as it has been replaced with Auth0 ProtectedRoute

**Comment by allraec** [comment]
Hello! Can we confirm with someone more authorized if all checks need to pass first? There seem to be one that is failing but I'm not entirely sure what it entails. Good job on the work transitioning from Okta to Auth0!

**Comment by tinomen** [comment]
Yes, all checks need to pass. There are 2 failing tests.

**Comment by Antashma** [comment]
The tests files for LoginContainer and ParentDashboardContainer seem to be the the issue for the test fail.  It looks like you have to edit these to test the Auth0 updates you made in those components.  Please update these test files!

**Comment by da-vazquez** [comment]
Updates to all original PR suggestions pushed and all failed tests are now passing. Thank you all for the feedback

**Review by kwnie — COMMENTED** [review]
Looks great, Daniel. All of your code is concise and clean.

**Review by kdneldor — COMMENTED** [review]
This PR is starting to look really good Daniel. We just need to clear up the concerns around these conversation items to get it ready for merge.

**Review by Antashma — CHANGES_REQUESTED** [review]

**Comment by da-vazquez** [comment]
Removed commented out code and test's are failing again, looking into root cause

**Review by kwnie — DISMISSED** [review]

File: src/index.js [hunk 8]
```diff
@@ -76,18 +74,26 @@ import { AudioBook } from './components/pages/AudioBook';
...
+
 ReactDOM.render(
   //
   <Router>
     <React.StrictMode>
       <Provider store={store}>
         <PersistGate persistor={persistor}>
-          <App />
+          <Auth0Provider
+            domain="dev-ww9ta3gj.us.auth0.com"
```
[the remark this answers is not in the document]
Comment (trevorjamesmartin) [reply src/index.js]: good catch!

**Review by kwnie — DISMISSED** [review]

File: sh.exe.stackdump [hunk 9]
```diff
@@ -0,0 +1,31 @@
+Stack trace:
+Frame        Function    Args
```
Comment (tinomen) [inline sh.exe.stackdump]: Let's remove file from the branch.

**Review by tinomen — CHANGES_REQUESTED** [review]
Lots of work to make this look simple. Well done.

**Review by kdneldor — DISMISSED** [review]
This looks good Daniel! The sh.exe.stackdump file needs to be deleted, and we need to coordinate the merge with the BE, but other than that it looks ready to go.

**Review by Antashma — DISMISSED** [review]

**Review by kwnie — DISMISSED** [review]

**Review by tinomen — DISMISSED** [review]

**Review by kdneldor — APPROVED** [review]
Great job switching out ParentDashboard.js with NewParentDashboard.js! I'm excited to coordinate with the BE PR to get this merged!

File: src/__tests__/ParentDashboardContainer.test.js [hunk 10]
```diff
@@ -5,35 +5,38 @@ import { Provider } from 'react-redux';
...
+  Auth0Provider: ({ children }) => children,
+  withAuthenticationRequired: (component, _) => component,
+  useAuth0: () => {
     return {
-      authState: {
-        isAuthenticated: true,
-      },
-      authService: {
-        getUser: () => Promise.reject(),
+      isAuthenticated: false,
```
Comment (tonysorensen) [inline src/__tests__/ParentDashboardContainer.test.js:19]: I'm just curious about the change in state in testing. Previously the test was done with isAuthenticated being true and now it is false. Why the change?

File: src/__tests__/RenderParentDashboard.test.js [hunk 11]
```diff
@@ -12,13 +12,18 @@ afterEach(() => {
...
-jest.mock('@okta/okta-react', () => ({
-  useOktaAuth: () => {
+jest.mock('@auth0/auth0-react', () => ({
+  Auth0Provider: ({ children }) => children,
+  withAuthenticationRequired: (component, _) => component,
+  useAuth0: () => {
     return {
-      authState: {
-        isAuthenticated: true,
+      isAuthenticated: false,
```
Comment (tonysorensen) [inline src/__tests__/RenderParentDashboard.test.js:20]: Also in this test, the is Authenticated boolean has changed. See comment in ParentDashboardContainer.test.js

File: src/api/index.js [hunk 12]
```diff
@@ -56,22 +60,22 @@ const getDSData = (url, authState) => {
     .catch(err => err);
 };
 
-const apiAuthGet = (endpoint, authHeader) => {
-  return axios.get(`${getApiUrl()}${endpoint}`, { headers: authHeader });
+const apiAuthGet = endpoint => {
+  return axios.get(`${getApiUrl()}${endpoint}`, getAuthHeader());
```
Comment (tonysorensen) [inline src/api/index.js:64]: I like the changes here where a middleware function is being passed to take care of the state being passed. My question is, on line 53 it looks like getAuthHeader() was assigned to headers. Is that necessary if we are passing in the actual function here? Can we just pass the variable 'headers'?

File: package.json [hunk 13]
```diff
@@ -8,6 +8,8 @@
     "@ant-design/icons": "^4.2.2",
     "@auth0/auth0-react": "^1.6.0",
     "@craco/craco": "^5.9.0",
+    "@okta/okta-react": "^3.0.2",
+    "@okta/okta-signin-widget": "^4.1.2",
```
Comment (Antashma) [inline package.json:12]: Glad you caught that!  That will save on future errors until the merge is  complete.

**Review by Antashma — APPROVED** [review]

**Review by kwnie — APPROVED** [review]
